source-academy/frontend
# CSE Machine: Enable visualization when running programs with runtime errors

### Description

This PR makes changes to the CSE machine visualization:

- The visualization is enabled for programs with runtime errors.  
- Adjusting the slider to step `i` now runs the program for `i` evaluation steps (previously the program was evaluated up to before step `i`.)
- The slider now goes from 0 to `min(number of steps needed to evaluate the program, step limit)` for programs evaluated without errors, or 0 to `min(number of steps evaluated before program runs into error + 1, step limit)` for programs with runtime errors.
- Adjusting the slider to `number of steps evaluated before program runs into error + 1` shows a popup with the error output, as shown in the REPL.

Implements and resolves #2854.

### Type of change


### How to test

1. Open Playground.
2. Insert the following code:
```js
1;
2;
3;
a;
const a = 1;
```
3. Go to the CSE machine tab.
4. Click `Run`. The visualization should look like this:
![](https://cdn.discordapp.com/attachments/910376383178502204/1219138658494582814/image.png?ex=660a365a&is=65f7c15a&hm=728d9bc768ad4bb255cf14c59105f7bef10fd90a093ae9e4dc11c750e4f6a91c&)
5. Step through the visualization. When you reach step 9, the visualization should look like this:
![](https://cdn.discordapp.com/attachments/910376383178502204/1219139238159847505/image.png?ex=660a36e4&is=65f7c1e4&hm=459daec09e6725d02996481f96ab750f32c690b9656036ef17b2364a5f9171ae&)

### Checklist
